packages/go/bomenc/utf16.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2024 Specter Ops, Inc.
+//
+// Licensed under the Apache License, Version 2.0
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package bomenc
+
+import (
+	"bytes"
+	"errors"
+)
+
+// utf16ToUTF8 converts a UTF-16 encoded byte slice to UTF-8.
+// It handles both big-endian and little-endian encodings, as well as surrogate pairs.
+//
+// Parameters:
+//   - data: A byte slice containing UTF-16 encoded data
+//   - bigEndian: A boolean indicating whether the input is big-endian (true) or little-endian (false)
+//
+// Returns:
+//   - A byte slice containing the UTF-8 encoded result
+//   - An error if the conversion fails (e.g., due to incomplete input)
+//
+// Advantages of using bitwise operations for this conversion:
+//
+//  1. Efficiency: Bitwise operations are extremely fast at the CPU level.
+//     They typically execute in a single clock cycle, making them more
+//     efficient than arithmetic operations or function calls.
+//
+//  2. Direct memory manipulation: Bitwise operations allow us to work
+//     directly with the binary representation of the data, which is
+//     crucial when dealing with different byte-order encodings.
+//
+//  3. Preserving data integrity: By using bitwise operations, we ensure
+//     that we're interpreting the bytes exactly as they are, without any
+//     unintended modifications that could occur with higher-level operations.
+//
+//  4. Endianness handling: Bitwise operations make it easy to handle both
+//     big-endian and little-endian encodings with minimal code duplication.
+//
+//  5. Performance in loops: When processing large amounts of text, the
+//     performance benefits of bitwise operations become significant due
+//     to the number of times these operations are repeated.
+//
+//  6. Low-level control: Bitwise operations provide fine-grained control
+//     over individual bits, which is necessary for correct interpretation
+//     of multi-byte character encodings.
+func utf16ToUTF8(data []byte, bigEndian bool) ([]byte, error) {
```

Comment:
There are a lot of bitwise operations and other arcane components in this function. Is there truly no upstream stdlib or library function that does the same so we don't need to maintain this?

See: https://pkg.go.dev/unicode/utf16